### ISSUE TITLE
Need a better way to handle GCC intrinsics

### ISSUE THREAD

**Issue by madscientist** [clangd/clangd#1695]
I've seen the various other issues which are marked closed and I get that there's not a lot of enthusiasm for tackling this issue, in clangd.  Unfortunately none of the various workarounds like issue #1203 or issue #1432 or issue #1345 are complete and they cannot solve the full problem.  In my situation I'm cross-compiling and I _have_ to use **--query-driver**.  And also my code needs to include intrinsics and it also uses third party code that includes intrinsics; suppressing diagnostics is not sufficient because clangd just refuses to parse these files, with lots of various errors.

In an ideal world clangd would figure this all out for itself and replace the directory containing the intrinsics in the database list that it gets by querying the driver with its own.  However, I understand this may be tricky.

Another option that might be sufficient for some people (for example, I could make it work) would be if the compiler driver had some way of knowing that it was being invoked by clangd.  Then for example, we could create (or modify, because often we already have one when we're cross-compiling with an alternative sysroot etc.) a wrapper around the compiler that would DTRT when it was invoked by clangd to obtain the compilation database.

As an example, if clangd would create and export an environment variable like `CLANGD_RESOURCE_DIR` which would be set to the path of clangd's resource directory when it invokes the compiler to retrieve the database, then I could create (or modify) a wrapper around my compiler which would replace the GCC intrinsics directory with the clangd resource directory (plus, I suppose, the `/include`).  I tried just hardcoding a bunch of stuff in my wrapper (obviously not portable even to other devs on my team) to test this, and it did work.

It's a bit clunky but it's very simple (I assume) to implement in clangd and it would give people who have more extreme environments at least the basic tools they need to implement their own solution, and give the clangd devs something to point to as a DIY facility.  With clangd's current feature set it's simply impossible to do in any reasonable way.

**Comment by HighCommander4** [comment]
> With clangd's current feature set it's simply impossible to do in any reasonable way.

Just for completeness, have you considered specifying the desired combination of built-in include paths manually?

That is, something like:

```yaml
CompileFlags:
  Add: [
    -nostdinc
    -I/cross/compiler/builtin/include/path/1
    -I/cross/compiler/builtin/include/path/2
    ... (skip the one you want replaced with clang's resource path)
    -I/clang/resource/path
  ]
```

**Comment by madscientist** [comment]
Unfortunately the cross compiler and the complete sysroot environment is checked into a tools Git repository, versioned, branched for old releases, and checked out by the user or CI/CD system into essentially random locations.  So no solution that relies on hardcoded paths to the compiler installation can work (for me anyway) :-(

**Comment by HighCommander4** [comment]
> As an example, if clangd would create and export an environment variable like `CLANGD_RESOURCE_DIR` which would be set to the path of clangd's resource directory when it invokes the compiler to retrieve the database, then I could create (or modify) a wrapper around my compiler which would replace the GCC intrinsics directory with the clangd resource directory (plus, I suppose, the `/include`).

If you're going to write such a wrapper which replaces the GCC intrinsics directory with the clang one, couldn't you just invoke that wrapper in a pre-build step, and have it generate a `.clangd` file along the lines in my previous comment?

**Comment by HighCommander4** [comment]
Does the issue of needing to run a portion of the build system prior to editing not also apply to `compile_commands.json` as well (which is presumably generated by the build system, and contains project-specific flags important for correctly parsing the code)?

**Comment by madscientist** [comment]
That is true, although of course not everyone uses `compile_commands.json` and it's not necessary.  However my other point about not have a way to obtain the resource directory path stands.

**Comment by HighCommander4** [comment]
I'd be curious to know more about what you use instead of `compile_commands.json`.

Requiring part of the build (enough to generate `compile_commands.json`, and often also to produce some generated header files) to run before clangd is usable is a constraint that I regularly suffer from as well; if you know a way around it, I'd like to learn about it.

---

Anyways, I have no opposition to your suggestion, I was just trying to brainstorm things that might already work for you today. Would you like to prototype a patch implementing it? I'm happy to give some pointers to the relevant code.

**Comment by madscientist** [comment]
You can avoid something like `compile_commands.json` IFF your build is simple enough that a standard set of compile line options is sufficient to build all the source (that is all the code compiles using the same options).  Depending on how you configure your system this can actually be not that difficult, especially insofar as LSP is concerned because mostly they just care about the preprocessor options and not all the other fancy options.  But, clearly depending on your build system it may not be possible.

I created a pull request for a change to add the environment variable.  This code only sets the variable for the actual invocation of the query driver, then unsets it again.  I'm confident it probably doesn't follow various requirements: I don't know if llvm has any kind of RAII facility like SCOPE_EXIT that might be preferred, probably there's someplace better to define the name of the environment variable, etc.

Also, definitely new docs will need to be written for clangd users to understand what to do with this variable and how it helps.

**Comment by HighCommander4** [comment]
Rather than trying to add the variable to clangd's own environment and letting the child process inherit it from there, I would suggest adding it only to the child process's environment; `llvm::sys::ExecuteAndWait()` has an `Env` parameter which can be used for this purposes.

Touching clangd's own environment is potentially problematic because it's a process-global piece of state, and clangd can parse (including running the system include extractor) multiple files in parallel.

**Comment by HighCommander4** [comment]
(Alternatively, we could add it to clangd's environment on startup, before clangd has spawned any worker threads.)

**Comment by madscientist** [comment]
I thought about adding it to the environment globally but decided to be cautious.  However I'm happy to add it globally.  The only question is where it can be added which is both global AND is late enough that we can retrieve the resource directory.

I had looked for an env parameter because that's clearly the best option (other than global) but I guess I was blind because now that I look again it's right there.  OK.

**Comment by madscientist** [comment]
Hm, it looks like if you pass an Env to ExecuteAndWait() it has to contain the ENTIRE environment, not just extra variables you want to have set.  That will be harder.

**Comment by madscientist** [comment]
I decided that it was simpler to set the variable globally; there doesn't seem to be any generic way to obtain the current process environment; many different tools appear to do it on their own.

I read a bit about phabricator but it says it's not maintained anymore since 2021.  Is this still the right way to manage patches? https://llvm.org/docs/Contributing.html#how-to-submit-a-patch

**Comment by HighCommander4** [comment]
> I read a bit about phabricator but it says it's not maintained anymore since 2021. Is this still the right way to manage patches? https://llvm.org/docs/Contributing.html#how-to-submit-a-patch

Yes, Phabricator is still what we use.

There is a plan to switch to GitHub PRs but my understanding is it won't happen until October or so.

**Comment by madscientist** [comment]
Hm, I pushed a review on phabricator https://reviews.llvm.org/D154903 but I wasn't able to add the right owner; CODE_OWNERS.TXT says:
```
N: Sam McCall
E: sammccall@google.com
D: clangd
```
but there's no phabricator handle here and my attempts to add the email address via cc etc. just gave errors.

Clearly there're tricks which are not specified in the documentation.  Hopefully someone will know what to do with this.

**Comment by sam-mccall** [comment]
> In an ideal world clangd would figure this all out for itself and replace the directory containing the intrinsics in the database list that it gets by querying the driver with its own. However, I understand this may be tricky.

Doh, I should have started reading from the beginning!

This is IMO the right answer, but replacement is not necessary: clang's builtin headers (resource dir) are already on the path. We just have to stop shadowing them with GCC etc's builtin headers.

https://reviews.llvm.org/D156044. I wonder if we should do path normalization in some cases (dotdots, windows)

**Comment by hefanbo** [comment]
[Exclude builtin headers from system include extraction](https://github.com/eymay/llvm-project/commit/5fbd71b62108fe5b048775451b3cfde5025af24f) results in the following problem:

1. Cross-compiling with *arm-gnu-toolchain*
2. `--query-driver=/opt/arm-gnu-toolchain/bin/arm-none-eabi-*`
3. In the following code, the wrong `stddef.h` file is referenced
   ```c
   #include <stdint.h> // Ctrl+click correctly opens /opt/arm-gnu-toolchain/arm-none-eabi/include/stdint.h
   #include <stddef.h> // Ctrl+click INCORRECTLY opens /usr/lib/llvm-18/lib/clang/18/include/stddef.h
   ```
4. Log of clangd
   ```
   System include extraction: target extracted: "arm-none-eabi"
   System include extraction: adding  /opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/include
   System include extraction: adding  /opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/include-fixed
   System include extraction: adding  /opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/../../../../arm-none-eabi/include
   System includes extractor: builtin headers /opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/include excluded
   System includes extractor: successfully executed /opt/arm-gnu-toolchain/bin/arm-none-eabi-gcc
   got includes: "/opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/include-fixed, /opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/../../../../arm-none-eabi/include"
   got target: "arm-none-eabi"
   ```
   The problem is `/opt/arm-gnu-toolchain/bin/../lib/gcc/arm-none-eabi/<version>/include` is found but excluded.

Referencing an incorrect header file is harmful:
1. The wrong header file may be incompatible with other files in the project, leading to false alarms.
2. The programmer can be misled when reviewing the code because he is reading the wrong file.

**Comment by HighCommander4** [comment]
> Referencing an incorrect header file is harmful

Of the two choices (1, using the build compiler's built-in headers, and. 2. using clang's built-in headers), it's not clear that one of them is "correct" and one of them is "incorrect".

Since the contents of the built-in headers are fairly closely coupled to the compiler front-end, and since clangd is parsing the code with the clang front-end, there is an argument for using clang's built-in headers.

In practice, it seems like some use cases work better if clang's built-in headers are used (e.g. code using compiler intrinsics, see the bug reports linked in the issue description), and other use cases may work better if the build compiler's built-in headers are used (e.g. for the reasons you mentioned).

Whichever we choose as the default, some users will prefer the other behaviour and need to take manual steps to achieve it.

(Maybe we could make it easier to choose the other behaviour, e.g. have an `--exclude-builtin-headers` command line flag that modifies the behaviour of `--query-driver`?)

**Comment by hefanbo** [comment]
Thanks for the clear explaination.

I still believe it is not proper for ctrl-click to point to a different file other than the one used at compile time. So it seems to me that replacing the built-in headers is a workaround to deal with problems such as gcc intrinsics.

> (Maybe we could make it easier to choose the other behaviour, e.g. have an `--exclude-builtin-headers` command line flag that modifies the behaviour of `--query-driver`?)

↑ This suggestion sounds good, before we find a better way to deal with both situations.

**Comment by HighCommander4** [comment]
> > (Maybe we could make it easier to choose the other behaviour, e.g. have an `--exclude-builtin-headers` command line flag that modifies the behaviour of `--query-driver`?)
> 
> ↑ This suggestion sounds good, before we find a better way to deal with both situations.

Could you file a new issue for this please, so we don't forget?

**Comment by madscientist** [comment]
The only fully-capable solution is to not use clangd as an LSP server, if you are compiling with GCC.  You need to use a GCC-based LSP server.  Unfortunately one does not exist (although there have been conversations about creating one).

The intrinsics headers are, as mentioned above, very tightly coupled with the compiler.  They contain special compiler tricks and things that won't work with any other compiler: they're basically part of the "glue" that converts standard interfaces into low-level compiler interfaces.  It's just not going to work (in my experience) to run clangd but use GCC's intrinsics headers.

That being said, there IS an argument to be made that standard headers like stddef.h should not be considered "intrinsics" like this; I think it is quite possible that these can be implemented, in whole or in part, in a portable way that isn't not very compiler dependent.  In an ideal world, these types of headers would be split from true intrinsics headers.

But that would be a lot of work and churn in both the GCC and LLVM projects and I suspect no one is really excited about putting in this time and effort, just for what is in the grand scheme of things a pretty minor inconvenience.

I will be very interested to see whether a setup with this new "exclude" option provided is actually usable in practice, at all.  It's been a long time since I messed with this, since it's been working for a year or more, but from what I recall trying to use the GCC intrinsics with clangd was an utter fail.

**Comment by HighCommander4** [comment]
> The only fully-capable solution is to not use clangd as an LSP server, if you are compiling with GCC. You need to use a GCC-based LSP server. Unfortunately one does not exist (although there have been conversations about creating one).

I agree with this.

But that still leaves open the possibility that, of the two imperfect solutions currently available (clangd language server + clang built-in headers, and clangd language server + gcc built-in headers), one isn't strictly better than the other in all scenarios.

**Comment by temyurchenko** [comment]
> clang's builtin headers (resource dir) are already on the path

If my reading of `Tooling::injectResourceDir` is correct, it's not the case. The resource dir is determined based on the binary location. In some cases, it leads to builtin headers being absent. Especially, when you are making another tool based on clangd.

So, in effect, the current solution excludes the builtin headers from the search and breaks compilation.

**Comment by HighCommander4** [comment]
> If my reading of `Tooling::injectResourceDir` is correct, it's not the case. The resource dir is determined based on the binary location. In some cases, it leads to builtin headers being not found. Especially, when you are making another tool based on clangd.
> 
> So, in effect, the current solution excludes the builtin headers from the search and breaks compilation.

In such situations, the builtin header directory can be added manually with [`CompileFlags`](https://clangd.llvm.org/config.html#add). (And once https://github.com/clangd/clangd/issues/2074 is fixed, it can then be done more easily by passing `--exclude-builtin-headers=false` or whatever the option ends up being called.)

**Comment by leonardopsantos** [comment]
> > With clangd's current feature set it's simply impossible to do in any reasonable way.
> 
> Just for completeness, have you considered specifying the desired combination of built-in include paths manually?
> 
> That is, something like:
> 
> CompileFlags:
>   Add: [
>     -nostdinc
>     -I/cross/compiler/builtin/include/path/1
>     -I/cross/compiler/builtin/include/path/2
>     ... (skip the one you want replaced with clang's resource path)
>     -I/clang/resource/path
>   ]

Yes, and you can get CMake to generate that for you at configure-time. Because of a quirk in my build environment, clangd wouldn't find glaze's headers, so I did this:

`CMakeLists.txt`
```cmake
include(FetchContent)
FetchContent_Declare(
  glaze
  GIT_REPOSITORY https://github.com/stephenberry/glaze.git
  GIT_TAG v6.0.1
  GIT_SHALLOW TRUE
)
FetchContent_MakeAvailable(glaze)

...

set(GLAZE_INCLUDE_DIR "${glaze_SOURCE_DIR}/include")
file(RELATIVE_PATH CLANGD_TEMPLATE_PATH "${CMAKE_SOURCE_DIR}" "${CMAKE_CURRENT_SOURCE_DIR}/clangd.in")
configure_file(
    "${CMAKE_CURRENT_SOURCE_DIR}/clangd.in"
    "${CMAKE_SOURCE_DIR}/.clangd"
    @ONLY
)
```

`clangd.in`
```
# Generated by CMake from @CLANGD_TEMPLATE_PATH@ — do not edit manually.
CompileFlags:
  Add: [
    "-isystem@GLAZE_INCLUDE_DIR@"
  ]

Diagnostics:
  Suppress: [builtin_definition]
```


The result is:

```
# Generated by CMake from src/something/something/clangd.in — do not edit manually.
CompileFlags:
  Add: [
    "-isystem/net/somewhere/somewhere/somewherelse/build/_deps/glaze-src/include"
  ]

Diagnostics:
  Suppress: [builtin_definition]
```

@madscientist 

> Unfortunately the cross compiler and the complete sysroot environment is checked into a tools Git repository, versioned, branched for old releases, and checked out by the user or CI/CD system into essentially random locations. So no solution that relies on hardcoded paths to the compiler installation can work (for me anyway) :-(

Maybe using CMake works for you.
